lib/modules.nix

```diff
@@ -615,7 +615,13 @@ let
           loc = prefix ++ [name];
           defns = defnsByName.${name} or [];
           defns' = defnsByName'.${name} or [];
-          optionDecls = filter (m: isOption m.options) decls;
+          optionDecls = filter
+            (m: m.options?_type
+                && (m.options._type == "option"
+                    || throwDeclarationTypeError loc m.options._type
```

Comment:
I'm a bit worried about the code complexity. Could this maybe be implemented here instead? https://github.com/NixOS/nixpkgs/blob/d45e075d11b02f7dee591a8d2599ba79e6b15674/lib/modules.nix#L566-L579

That code would've thrown the error beforehand

Reply:
`byName` is also used for definitions, not just declarations, oddly enough. It's an example of DRY gone wrong.
If anything, we should get rid of it - inline it.
Why would that place make it less complex?

Reply:
To me it seems like there's two code paths handling the same kind of error right now.

Reply:
Agreed that `byName` is DRY gone wrong btw!

Reply:
Let's fix that first:
 - #242812